### ISSUE TITLE
FedX runs into Phaser > 65535

### ISSUE THREAD

**Issue by qootec** [eclipse/rdf4j#4610]
### Current Behavior

Perform "simple" query on the FedX (see below for more info)
```
select * where { 
    ?x a :X;
       :value ?value;
       :relation1 ?p1.

    ?y a :Y;
       :relation2 ?p1;
       :relation1 ?p2.
    
    ?z a :Z;
       :relation3 ?p2.
}
```

Result is: 
```
Query evaluation error: java.lang.IllegalStateException: Attempt to register more than 65535 parties for java.util.concurrent.Phaser@1707c5f1[phase = 0 parties = 65535 arrived = 0] (HTTP status 500)
```

### Expected Behavior

Normal query response.
The following reduction on the query's complexity avoids the issue:
```
select * where { 
    ?x a :X;
       :value ?value;
       :relation1 ?p1.

    ?y a :Y;
       :relation2 ?p1;
       :relation1 ?p2.
}
```

### Steps To Reproduce

Environment:
* GraphDB 10.2.1, using RDF4j 4.2.3
* FedX repo consisting of two small repos (number of triples between ())
  - Repo1: Local GraphDB repo (10946)
  - Repo2: Remote SPARQL repo pointing to an Ontop repo (1968)

Repo contents:
* Repo1: Concepts X and Z and the relations related to them
* Repo2: Concept Y and the relations related to them
* X, Y, Z all inherit from a parent Concept.

Performed the query in the GraphDB query window.

### Version

4.2.3

### Are you interested in contributing a solution yourself?

None

### Anything else?

* I can provide a reproduction environment if desired.
* Looks like #3106 , but that one is quite old and seems to be fixed.

**Comment by aschwarte10** [comment]
Thanks for filing up this issue and the example scenario. I also thought we configured the phaser correctly (to use the nesting of phasers) with https://github.com/eclipse/rdf4j/issues/3106

Maybe some case there was missed or something else changed meanwhile.

I will try to look into this and come up with a reproducible test environment. However, I would need a bit of time as I am quite busy these days. Will come back here once I have something

**Comment by qootec** [comment]
Thanks for your effort.
Two additional points for your information:
* The problem is not occurring when I re-order the statements in the first query as follows:
```
select * where { 
    ?y a :Y;
       :relation2 ?p1;
       :relation1 ?p2.
    
    ?x a :X;
       :value ?value;
       :relation1 ?p1.

    ?z a :Z;
       :relation3 ?p2.
}
```
* If you want a reproducible environment, I can give you (in private) an archive of the contents and definition of my repositories. I guess that would automatically result in a reproducible context.

**Comment by aschwarte10** [comment]
I found a place in the code of one of the join handlers, where the registration to the phaser was wrong. This should fix the issue: https://github.com/eclipse/rdf4j/pull/4613

Note: I had a local test environment (with some debug tweaking) where I was able to reproduce the error before, and where it was gone after. Unfortunately this cnanot be provided as unit test as it very much depends on concurrency.

**Comment by qootec** [comment]
Thanks Andreas for this fix.

Although I would obviously like to try the fix, I do not immediately see how I could: I'm using rdf4j through its integration in GraphDB, so I guess I'll need to wait till Ontotext releases a new GraphDB that integrates your fix, no?

**Comment by hmottestad** [comment]
I've released this in 4.3.2. It **might** make it into the next minor release of GraphDB (10.3).
